MSMAll/scripts/SingleSubjectConcat.sh

```diff
@@ -31,504 +31,171 @@
-		
-	done
-	mkdir -p "${OutputFolder}"
-	${Caret7_Command} -cifti-merge ${OutputFolder}/${OutputfMRIName}${fMRIProcSTRING}${OutputProcSTRING}.dtseries.nii ${MergeSTRING}
-}
-
-# ------------------------------------------------------------------------------
-#  "Global" processing - everything above here should be in a function
-# ------------------------------------------------------------------------------
-
-# Establish defaults
-G_DEFAULT_MATLAB_RUN_MODE=1		# Use interpreted MATLAB
-
-# Set global variables
-g_script_name=$(basename "${0}")
-
-# Allow script to return a Usage statement, before any other output
-if [ "$#" = "0" ]; then
-    show_usage
-    exit 1
+done
+if [[ "$EndFrame" == "" ]]; then
+	EndFrame="$minFrames"
+	log_Msg "EndFrame: ${EndFrame}"
 fi
-
-# Verify that HCPPIPEDIR environment variable is set
-if [ -z "${HCPPIPEDIR}" ]; then
-	echo "${g_script_name}: ABORTING: HCPPIPEDIR environment variable must be set"
-	exit 1
+# check valid frame range
+if [ $((StartFrame)) -gt "${minFrames}" ] || [ $((EndFrame)) -gt "${minFrames}" ] || [ $((StartFrame)) -lt 1 ] || [ $((EndFrame)) -lt 1 ] || [ $((StartFrame)) -gt $((EndFrame)) ]; then
+	log_Err_Abort "The provided start frame ${StartFrame} and end frame ${EndFrame} is not valid, it must be between the range 1 to ${minFrames}. Please check the fMRI runs."
 fi
+log_Msg "StartFrame: ${StartFrame}, EndFrame: ${EndFrame}"
+FrameString="Frame${StartFrame}To${EndFrame}"
+duration=0
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index))
+do
+	TR="${runTR[index]}"
+	Frames=$((EndFrame-StartFrame+1))
+	duration=$(bc -l <<< $duration+$TR*$Frames)
+done
+
+# convert to mins secs
+m=$(bc <<< "${duration}/60")
+s=$(bc <<< "${duration}%60")
+DurationString="${m}mins${s}secs"
+
+MergeArray=()
+cnt=0
+# loop for demean+vn+concat
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index)) ; do
+    fMRIName="${fMRINamesArray[$index]}"
+	log_Msg "fMRIName: ${fMRIName}"
+	ResultsFolder="${AtlasFolder}/Results/${fMRIName}"
+	log_Msg "ResultsFolder: ${ResultsFolder}"
+	
+	FullDenseTCS=${ResultsFolder}/${fMRIName}${fMRIProcSTRING}.dtseries.nii
+	# temporary files
+	FrameDenseTCS=${ResultsFolder}/${fMRIName}${fMRIProcSTRING}_${FrameString}.dtseries.nii
+	FrameMean=${ResultsFolder}/${fMRIName}${fMRIProcSTRING}_${FrameString}_mean.dscalar.nii
+	FrameOutput=${ResultsFolder}/${fMRIName}${fMRIProcSTRING}_${FrameString}${OutputProcSTRING}.dtseries.nii
+
+	# pick frames
+	${Caret7_Command} -cifti-merge ${FrameDenseTCS} -cifti ${FullDenseTCS} -column ${StartFrame} -up-to ${EndFrame}
+
+	# mean file
+	${Caret7_Command} -cifti-reduce ${FrameDenseTCS} MEAN ${FrameMean}
+	MATHDemean=" - Mean"
+	VarDemean="-var Mean ${FrameMean} -select 1 1 -repeat"
+	
+	# vn file
+	OutputVN="${ResultsFolder}/${fMRIName}${fMRIProcSTRING}_vn.dscalar.nii"
+	log_File_Must_Exist "$OutputVN"
+	MATHVN=" / max(VN,0.001)"
+	VarVN="-var VN ${OutputVN} -select 1 1 -repeat"
+	
+  	# math expression
+	MATH="(TCS${MATHDemean})${MATHVN}"
```

Comment:
If we aren't going to have conditionals controlling whether we do the demean and variance normalize, please just have the final `"TCS - Mean / max(VN, 0.001)"` expression as a literal in the -cifti-math line itself, rather than splitting it across multiple variables.